Add a websocket route to the Internal HTTP API

rjobanp opened the pull request:
### Motivation

  * This PR adds a known-desirable feature: Fixes: https://github.com/MaterializeInc/materialize/issues/22303

See issue for details ^

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

The auth changes to websocket handling should only come into play when used in the InternalAPIServer which itself attaches the `internal_http_auth` middleware, but would appreciate eyes double-checking that: https://github.com/MaterializeInc/materialize/blob/856f69d03272e1eab6e32f2dec339d83eb7dcf20/src/environmentd/src/http.rs#L460

I also noticed that the default for that middleware is the system user, not the support user, which contradicts the issue written by @benesch , but I think is consistent with the previous behavior before the header existed. Looking for opinions on whether I should change this to mz_support, and what might break if I do.

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

rjobanp commented:
@RobinClowers successfully tested in my personal stack :) 

<img width="1344" alt="image" src="https://github.com/MaterializeInc/materialize/assets/3818834/42179c49-3f26-4ea8-bfb4-681e1bea51ac">

it looks like the auth refresh isn't properly working since I get kicked out after some time, but it's pretty usable for basic impersonation!

RobinClowers commented:
> it looks like the auth refresh isn't properly working since I get kicked out after some time, but it's pretty usable for basic impersonation!

Yeah, this is a known issue, because the frontegg cookie won't work on the teleport domain. Good news, I'm getting rid of the frontegg login requirement for impersonation completely!

RobinClowers reviewed (approved):
Nice work! It definitely would have taken me more than an afternoon to knock this out, thanks for taking it on.

mjibson commented on src/environmentd/src/http.rs:
```diff
@@ -726,16 +746,31 @@ async fn init_ws(
+            }
+        };
+        (auth(frontegg, creds).await?, options)
+    } else if let Some(user) = existing_user {
```
This 4 branch if block is now really hard to reason about, especially because it depends on the state of 3 things (frontegg, ws_auth, existing_user), which makes it very easy to miss states and figure out if they should be problematic or not. For example, what happens if both frontegg and existing_user are specified? That should be impossible right? The block here doesn't complain if so. This needs to be changed to a match statement to it is exhaustive.

rjobanp replied:
Great point -- this got pretty hairy and I should have refactored it, your version looks a lot cleaner

mjibson commented on src/environmentd/src/http.rs:
```diff
@@ -726,16 +746,31 @@ async fn init_ws(
+        match ws_auth {
+            WebSocketAuth::OptionsOnly { options }
+            | WebSocketAuth::Basic {
+                user: _,
```
When is this ever possible? If we never expect basic auth AND a passed user header, this should error.

rjobanp replied:
This case allows the test to use `util::auth_with_ws(&mut ws, options).unwrap();` which uses basic auth; so similar to the console case I opted to ignore the other fields and just pull the `options` field if the user header is provided to avoid major code changes there.

mjibson commented on src/environmentd/src/http.rs:
```diff
@@ -726,16 +746,31 @@ async fn init_ws(
+                password: _,
+                options,
+            }
+            | WebSocketAuth::Bearer { token: _, options } => (user.clone(), options),
```
The bearer case also seems like it should error because we never expect it.

rjobanp replied:
I added this in to avoid any frontend changes in the web console, since the frontend can still send a token along with it's initial auth message and we will just ignore it if the user is already set. That's what allowed me to be able to test this e2e without any console changes.

Happy to go either way on this -- if we assume the console will have a conditional around the auth logic when in user impersonation mode then being stricter here is fine

mjibson reviewed (commented):
I pushed a second commit with some ideas about other ways to improve the clarity of the auth method. Thoughts?

benesch commented on src/environmentd/src/http.rs:
```diff
@@ -439,9 +439,22 @@ impl InternalHttpServer {
+            .layer(Extension(Arc::clone(&active_connection_count)));
+
+        let ws_router = Router::new()
+            .route("/api/experimental/sql", routing::get(sql::handle_sql_ws))
```
But we didn't do that for the /api/sql URL! IMO this is morally equivalent to the public API and so it makes sense to use  the same path as the public API to indicate the correspondence. Also it's one less conditional in the console codebase.

rjobanp replied:
Yep I left this the same so that there would be no changes needed on the web console - it can just use the same code without a conditional (and that's how I was able to test that it worked)

benesch commented on src/environmentd/src/http.rs:
```diff
@@ -709,33 +729,57 @@ async fn init_ws(
+        }
+        // No frontegg, specified existing user, we do not expect basic or bearer auth.
+        (None, Some(_), WebSocketAuth::Basic { .. } | WebSocketAuth::Bearer { .. }) => {
+            anyhow::bail!("unexpected")
```
Can we log a more informative error message here?

benesch commented on src/environmentd/src/http.rs:
```diff
@@ -709,33 +729,57 @@ async fn init_ws(
+        (Some(_), Some(_), _) => anyhow::bail!("unexpected"),
+        // No frontegg, no existing user, and no passed username.
+        (None, None, WebSocketAuth::Bearer { .. } | WebSocketAuth::OptionsOnly { .. }) => {
+            anyhow::bail!("unexpected")
```
Ditto.

benesch reviewed (commented):
We should have @matthewarthur do a security review of this as well.

matthewarthur commented:
> ... middleware is the system user, not the support user, which contradicts the issue written by @benesch , but I think is consistent with the previous behavior before the header existed. Looking for opinions on whether I should change this to mz_support, and what might break if I do.

Looks like internal_http_auth is only used for tests atm? I like changing the default to ms_support.

matthewarthur commented on src/environmentd/src/http.rs:
```diff
@@ -709,33 +729,57 @@ async fn init_ws(
     };
-    let (creds, options) = if frontegg.is_some() {
-        match ws_auth {
+    let (user, options) = match (frontegg.as_ref(), existing_user, ws_auth) {
```
@RobinClowers all frontegg involvement is going away? really interested in that, please tag me when you put that code up

matthewarthur reviewed (approved):
+1 to thanks for knocking this out

benesch commented:
>> ... middleware is the system user, not the support user, which contradicts the issue written by @benesch , but I think is consistent with the previous behavior before the header existed. Looking for opinions on whether I should change this to mz_support, and what might break if I do.
> Looks like internal_http_auth is only used for tests atm? I like changing the default to ms_support.

The internal HTTP server is also used by the Prometheus SQL exporter and the billing exporter in cloud. I'm 👍🏽 on changing the default here, but both of those services rely on the default being `mz_system`. So we should do that change in a dedicated follow up PR.

rjobanp commented:
> I pushed a second commit with some ideas about other ways to improve the clarity of the auth method. Thoughts?

@mjibson I definitely prefer the match statement you introduced - thank you for that.

I left a couple of responses to your initial feedback -- we need to decide if we want to reject all 'unexpected' cases and require a conditional in the web console, or ignore the token/user/password fields when handling the internal-api auth and allow the console to work as-is without code changes. Thoughts on that?

mjibson commented:
> we need to decide if we want to reject all 'unexpected' cases and require a conditional in the web console, or ignore the token/user/password fields

Let's say the passed user in the header differs from the one presented in the initial ws auth message. Which one should we use? Should we verify that the presented auth credentials are valid, but then ignore the user for them?

rjobanp commented:
> Let's say the passed user in the header differs from the one presented in the initial ws auth message. Which one should we use? Should we verify that the presented auth credentials are valid, but then ignore the user for them?

That's a good point -- the way I had it implemented it would just ignore the auth credentials entirely and wouldn't verify them (since the `frontegg` var was set to `None` for the Internal HTTP API WsState), so it'd always use the `x-materialize-user` header. In the case of user-impersonation we don't actually care what user is logged in to frontegg since they will likely be part of a different org-id entirely

RobinClowers commented:
> Let's say the passed user in the header differs from the one presented in the initial ws auth message. Which one should we use? Should we verify that the presented auth credentials are valid, but then ignore the user for them?

This case also exists for the regular http endpoint, clients can include an auth header with a token for some other user and it's just ignored. I vote we just ignore the token field on this internal interface.

That said, if we feel strongly that the token should not be allowed, it's an easy change on the console side.

mjibson commented:
> part of a different org-id entirely

Because we're doing things with authentication, employee special accounts, ignoring our auth provider in some cases, and (currently) providing employee credentials to user environments: I think it's worthwhile to do the extra work to make this a bit more paranoid. We should:

- Only allow OptionsOnly when accepting a user via header and error if any other credentials are even presented.
- Force the console to be aware it's in this mode and stop sending creds in that case. How hard is this? Is it a huge refactor? Can it be done automatically or will this require our employees click an additional button? Does the console know when it's in teleport mode?

mjibson commented:
> This case also exists for the regular http endpoint

Ah! I think we should change this too then. Sending ignored auth credentials that are valid for different environments can't be good.

rjobanp commented:
Sounds good, so let's go with the current approach on the PR and we'll just need to update Console to remove the `token` field when in impersonation mode.

I'll fix the test case to send OptionsOnly and then this PR should be ready for another review

rjobanp commented:
Okay I think I've addressed all comments, and updated the test case to check that we now error if bearer/token auth is provided to the internal websocket endpoint, but that providing "options" only works
